proposed/simple-csharp-programs-motivation-walkthrough.md

```diff
@@ -0,0 +1,311 @@
+
+It is also designed to be incremental. Not everything written here must be implemented. For example, we could ship a version without support for packages/frameworks.
+
+## Motivation
+
+C# has lots of great things going for it, but it falls short in a key area: getting started with the least possible friction.
+
+Lots of developers in the world have experience with languages like JavaScript, Python, and Ruby. In these languages, you typically get started in one of these three ways:
+
+1. Launch a REPL and start typing code and getting results
+2. Create a file with the right extension and run the code in that file
+3. Use build tooling to create an application and then run that application
+
+If you look at online guides for the above languages, they tend to start people off with (1) and (2) and rarely with (3).
+
+In the C# world, we start people off with (3) with the base tooling, the .NET CLI.
+
+This is problematic, because these people **do not expect that they need to do (3) to get started**. The tools don't support their instincts and there is very little information online to support them.
+
+Of course, REPLs are another great way to get started. But they are outside the scope of this document.
+
+## What is a Simple C# Program
+
+A Simple C# Program is this:
+
+* A C# file
+* A set of global `using`s inferred by the build environment
+* A tool that can run that file / produce a working program
+* A way to incorporate libraries and frameworks/sdks into that file and run it
+* A way to "graduate" to using a project file
+
+A Simple C# Program is **not** a replacement for project-based C# code or .NET Interactive.
+
+## Simple C# is just C# code
+
+Simple C# Programs are not a dialect of the C# language, like the interactive dialect, CSX. Any code in a Simple C# Program can be built and run in a project-based program.
+
+## User experience walkthrough
+
+Given the following code in `Hello.cs`:
+
+```csharp
+WriteLine("Hello World");
+```
+
+You can just run it:
+
+```bash
+$ dotnet run
+Hello World
+```
+
+Or with a shebang,
+
+```csharp
+#!/usr/bin/shared/dotnet
+WriteLine("Hello World");
+```
+
+and running:
+
+```bash
+$ ./hello.cs
+Hello World
+```
+
+The `using`s that are normally in a C# project template are implicitly available at design-time. This is accomplished with [Global Using Directives](https://github.com/dotnet/csharplang/blob/main/proposals/GlobalUsingDirective.md) that the build tooling will have available.
+
+The shebang is covered in [Shebang (#!) support](https://github.com/dotnet/csharplang/issues/3507).
+
+### Building instead of running
+
+You can also build it with `dotnet build`:
+
+```bash
+$ dotnet build
+$ ls
+Hello.exe
+$ ./Hello.exe
+Hello World
+```
+
+When built, it creates a single executable that it drops in the current working directory.
+
+### Referencing a package
```

Comment:
How was the need for NuGet packages in simple C# evaluated? From my understanding, the goal is to make C# more approachable. The `#r "nuget: X"` syntax is a special one-off feature that newbies will need to "unlearn" when graduating to a project file.

Will this "unlearning" process confuse new users? Would it be better to save NuGet support to project files only?

Reply:
There's upside and downside to doing/not doing it. Not necessarily the same audience, but for F# developers the `#r` nuget support has gone absolutely gangbusters, and ex-python programmers praise the simplicity. It's also supported in .NET Interactive and is the primary way to do things like chart a data set. But your points about learning two different ways is completely valid. Some degree of this may be necessary in the future. But this could be done incrementally as mentioned, so that package support comes later (if at all).

Reply:
I think this depends on the audience. Newcomers might not need NuGet packages, but I can see a lot of experienced C# developers using Simple Programs as a low-ceremony way to knock out ad-hoc programs and scripts. That latter audience will need+want package support.